src/v/container/include/container/contiguous_range_map.h

```diff
@@ -0,0 +1,298 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "container/fragmented_vector.h"
+
+/**
+ * Contiguous range map is an associative sorted container backed by
+ * chunked_vector designed to efficiently store objects indexed with contiguous
+ * and limited range of integers starting at 0. The container provides a
+ * wrapper around basic chunked vector that reassembles the API of a standard
+ * map.
+ *
+ * The wrapper allows random inserts of elements to the map although this is
+ * not supported by standard vector.
+ *
+ * Underlying container is resized every time its size is not sufficient to
+ * account for emplaced key.
+ *
+ * The contiguous_range_map tolerates gap in the range of keys however number
+ * and size of gaps is proportional to performance penalty hit when incrementing
+ * or decrementing iterators.
+ *
+ * NOTE:
+ * A map can store elements for key's which range starts from value greater than
+ * 0 but it will cause significant memory loss. If we will need to store such a
+ * ranges we may introduce non type template parameter indicating offset.
+ */
+template<std::integral KeyT, typename ValueT>
+class contiguous_range_map {
+public:
+    using value_type = std::pair<const KeyT, ValueT>;
+    using key_type = KeyT;
+    using mapped_type = ValueT;
+
+private:
+    using underlying_t = chunked_vector<std::optional<value_type>>;
+
+private:
+    template<bool Const>
+    class iter {
+    public:
+        using iterator_category = std::forward_iterator_tag;
+        using value_type =
+          typename std::conditional_t<Const, const value_type, value_type>;
+
+        using difference_type = std::ptrdiff_t;
+        using pointer = value_type*;
+        using reference = value_type&;
+
+        iter() = default;
+
+        /**
+         * Conversion operator allowing iterator to be converted to
+         * const_iterator, as required by the general iterator contract.
+         */
+        operator iter<true>() const { // NOLINT(hicpp-explicit-conversions)
+            iter<true> ret;
+            ret._container = _container;
+            ret._it = _it;
+            return ret;
+        }
+
+        reference operator*() const { return _it->value(); }
+
+        pointer operator->() const { return &(_it->value()); }
+
+        iter& operator++() {
+            ++_it;
+            skip_to_next_present();
+            return *this;
+        }
+
+        iter& operator--() {
```

Comment:
Why is this needed if the iterator is only a forward iterator?

Reply:
changed iterator_category to `std::bidirectional_iterator_tag`